File: hw/gongora/U2/HW21MONGODB/java/StudentsMongo/src/main/java/ec/edu/espe/studentsmongo/model/Menu.java
package ec.edu.espe.studentsmongo.model;

import java.util.Scanner;
import org.bson.Document;

/**
 *
 * @author Lucas Gongora
 */

public class Menu {
    private Scanner scanner;
    private CRUDMongo database;
    private boolean executionMenu;

    public Menu() {
        this.scanner = new Scanner(System.in);
        this.executionMenu = true;
        this.database = new CRUDMongo("mongodb+srv://lngongora:lngongora@cluster0.y73d4.mongodb.net/", "OOP");
    }

    public void initialize() {
        try {
            database.connectDB();
        } catch (Exception e) {
            System.err.println("Failed to initialize database connection.");
            System.exit(1);
        }
    }

    public String input(String question) {
        System.out.print(question);
        return scanner.nextLine();
    }

    public void run() {
        initialize();
        do {
            showOptions();
            String option = input("Enter the option: ");
            processOption(option);
        } while (executionMenu);

        scanner.close();
    }

    public void showOptions() {
        System.out.println("Students Menu");
        System.out.println("1. Add student");
        System.out.println("2. View all students");
        System.out.println("3. Update student");
        System.out.println("4. Delete student");
        System.out.println("5. Exit");
    }

    public void processOption(String option) {
        switch (option) {
            case "1":
                addStudent();
                break;
            case "2":
                viewAllStudents();
                break;
            case "3":
                updateStudent();
                break;
            case "4":
                deleteStudent();
                break;
            case "5":
                executionMenu = false;
                database.closeConnection();
                System.out.println("Exiting...");
                break;
            default:
                System.out.println("Invalid option, please try again.");
                break;
        }
    }

    public void addStudent() {
        String id = input("Enter the id: ");
        String name = input("Enter the name: ");
        double pga = Double.parseDouble(input("Enter the PGA: "));
        int age = Integer.parseInt(input("Enter the age in years: "));

        // Create JSON-like document structure
        Document json = new Document()
                .append("id", id)
                .append("name", name)
                .append("PGA", pga)
                .append("ageInYears", age);

        database.insertData("students", json);
    }

    public void viewAllStudents() {
        database.findData("students", new Document());
    }

    public void updateStudent() {
        String id = input("Enter the id of the student you want to update: ");
        Document query = new Document("id", id);
        database.findData("students", query);

        String name = input("Enter the new name: ");
        double pga = Double.parseDouble(input("Enter the new PGA: "));
        int age = Integer.parseInt(input("Enter the new age: "));

        Document updateDoc = new Document()
                .append("name", name)
                .append("PGA", pga)
                .append("ageInYears", age);

        database.updateData("students", query, updateDoc);
    }

    public void deleteStudent() {
        String id = input("Enter the id of the student you want to delete: ");
        Document query = new Document("id", id);
        database.findData("students", query);
        database.deleteData("students", query);
    }

    
}

File: hw/gongora/U2/HW21MONGODB/javascript/Menu.mjs
import readline from 'readline';
import { CRUDMongo } from './CRUDMongo.mjs';

export class Menu {
    #formConsole;
    #database;
    #executionMenu;

    constructor() {
        this.#formConsole = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        this.#executionMenu = true;
        this.#database = new CRUDMongo("mongodb+srv://lngongora:lngongora@cluster0.y73d4.mongodb.net/", "OOP");
    }

    async initialize() {
        try {
            await this.#database.connectDB();
        } catch (error) {
            console.error("Failed to initialize database connection.");
            process.exit(1);
        }
    }

    input(question) {
        return new Promise((resolve) => {
            this.#formConsole.question(question, answer => {
                resolve(answer);
            });
        });
    }

    async run() {
        await this.initialize();
        do {
            this.showOptions();
            let option = await this.input("Enter the option: ");
            await this.processOption(option);
        } while (this.#executionMenu);

        this.#formConsole.close();
    }

    showOptions() {
        console.log("Students Menu");
        console.log("1. Add student");
        console.log("2. View all students");
        console.log("3. Update student");
        console.log("4. Delete student");
        console.log("5. Exit");
    }

    async processOption(option) {
        switch (option) {
            case "1":
                await this.addStudent();
                break;
            case "2":
                await this.viewAllStudents();
                break;
            case "3":
                await this.updateStudent();
                break;
            case "4":
                await this.deleteStudent();
                break;
            case "5":
                this.#executionMenu = false;
                await this.#database.closeConnection();
                console.log("Exiting...");
                break;
            default:
                console.log("Invalid option, please try again.");
                break;
        }
    }

    async addStudent() {
        let id = await this.input("Enter the id: ");
        let name = await this.input("Enter the name: ");
        let PGA = parseFloat(await this.input("Enter the PGA: "));
        let ageInYears = parseInt(await this.input("Enter the age in years: "), 10);

        let json = {
            id: String(id),
            name: String(name),
            PGA: Number(PGA),
            ageInYears: Number(ageInYears)
        };

        try {
            await this.#database.insertData("students", json);
            console.log("Student added successfully!");
        } catch (err) {
            console.error("Error adding student:", err);
        }
    }

    async viewAllStudents() {
        try {
            let students = await this.#database.findData("students");
            if (students.length === 0) {
                console.log("No students found.");
            } else {
                console.log("Students List:");
                students.forEach(student => {
                    console.log(`ID: ${student.id}, Name: ${student.name}, PGA: ${student.PGA}, Age: ${student.ageInYears}`);
                });
            }
        } catch (err) {
            console.error("Error fetching students:", err);
        }
    }

    async updateStudent() {
        let id = await this.input("Enter the id of the student you want to update: ");
        let student = await this.#database.findData("students", { id });

        if (student.length === 0) {
            console.log("No student found with that ID.");
            return;
        }

        let name = await this.input(`Enter the new name (current: ${student[0].name}): `);
        let PGA = parseFloat(await this.input(`Enter the new PGA (current: ${student[0].PGA}): `));
        let ageInYears = parseInt(await this.input(`Enter the new age (current: ${student[0].ageInYears}): `), 10);

        let updateDoc = {
            name: String(name),
            PGA: Number(PGA),
            ageInYears: Number(ageInYears)
        };

        try {
            await this.#database.updateData("students", { id }, updateDoc);
            console.log("Student updated successfully!");
        } catch (err) {
            console.error("Error updating student:", err);
        }
    }

    async deleteStudent() {
        let id = await this.input("Enter the id of the student you want to delete: ");
        let student = await this.#database.findData("students", { id });

        if (student.length === 0) {
            console.log("No student found with that ID.");
            return;
        }

        try {
            await this.#database.deleteData("students", { id });
            console.log("Student deleted successfully!");
        } catch (err) {
            console.error("Error deleting student:", err);
        }
    }
}
